File: genetic-packer/src/main/java/genetic/packer/evolution/mutation/impl/MutatorImpl.java
package genetic.packer.evolution.mutation.impl;

import java.util.concurrent.ThreadLocalRandom;
import java.util.function.BiConsumer;
import java.util.function.Supplier;

import genetic.api.individual.Cell;
import genetic.api.individual.Organism;
import genetic.packer.evolution.mutation.Mutator;
import javafx.geometry.Bounds;
import javafx.scene.shape.Box;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Component;

/**
 * @author piotr.larysz
 */
@Component
public class MutatorImpl implements Mutator {

    private static final double mutationChance = 0.008;

    @Autowired
    @Qualifier("boundsAwareBoxTranslationRandomizer")
    private BiConsumer<Box, Bounds> translationRandomizer;

    @Autowired
    private Supplier<ThreadLocalRandom> random;

    private final Supplier<Boolean> mutation = () -> random.get().nextDouble() <= mutationChance; //todo make some component holding the random also

    @Override
    public void accept(Organism<Box> individual, Bounds bounds) {
        individual.getCells()
            .map(Cell::getNucleus)
            .forEach(box -> {
                if(mutation.get()) {
                    this.translationRandomizer.accept(box, bounds);
                }
            });
    }
}

File: genetic-packer/src/main/java/genetic/packer/evolution/mutation/impl/MutatorImpl.kt
package genetic.packer.evolution.mutation.impl

import genetic.api.individual.Organism
import genetic.api.mutation.Mutator
import javafx.geometry.Bounds
import javafx.scene.shape.Box
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Component
import java.util.concurrent.ThreadLocalRandom

@Component
class MutatorImpl(
    @Qualifier("boundsAwareBoxTranslationRandomizer") private val translationRandomizer: (Box, Bounds) -> Unit,
    private val random: () -> ThreadLocalRandom
) : Mutator<Bounds, Box> {

  private val mutation = {
    random().nextDouble() <= mutationChance
  }

  override fun invoke(individual: Organism<Box>, bounds: Bounds) {
    individual.cells
        .map { it.nucleus }
        .forEach {
          if (mutation()) {
            translationRandomizer(it, bounds)
          }
        }
  }

  companion object {
    private const val mutationChance = 0.008
  }
}